mesonbuild/msubprojects.py

```diff
@@ -0,0 +1,221 @@
+        update_file(wrap, repo_dir, options)
+    elif wrap.type == 'git':
+        update_git(wrap, repo_dir, options)
+    elif wrap.type == 'hg':
+        update_hg(wrap, repo_dir, options)
+    elif wrap.type == 'svn':
+        update_svn(wrap, repo_dir, options)
+    else:
+        mlog.log('  -> Cannot update', wrap.type, 'subproject')
+
+def checkout(wrap, repo_dir, options):
+    if wrap.type != 'git' or not os.path.isdir(repo_dir):
+        return
+    branch_name = options.branch_name if options.branch_name else wrap.get('revision')
+    mlog.log('Checkout %s in %s...' % (branch_name, wrap.name))
+    try:
+        git(['checkout', '-B', branch_name, '--'], repo_dir)
+        git_show(repo_dir)
+    except subprocess.CalledProcessError as e:
+        out = e.output.decode().strip()
+        mlog.log('  -> ', mlog.red(out))
+
+def download(wrap, repo_dir, options):
+    mlog.log('Download %s...' % wrap.name)
+    if os.path.isdir(repo_dir):
+        mlog.log('  -> Already downloaded')
+        return
+    try:
+        r = Resolver(os.path.dirname(repo_dir))
+        r.resolve(wrap.name)
+        mlog.log('  -> done')
+    except WrapException as e:
+        mlog.log('  ->', mlog.red(str(e)))
+
+def add_common_arguments(p):
+    p.add_argument('--sourcedir', default='.',
+                   help='Path to source directory')
+    p.add_argument('subprojects', nargs='*',
+                   help='List of subprojects (default: all)')
+
+def add_arguments(parser):
+    subparsers = parser.add_subparsers(title='Commands', dest='command')
+    subparsers.required = True
+
+    p = subparsers.add_parser('update', help='Update all subprojects from wrap files')
+    p.add_argument('--rebase', default=False, action='store_true',
+                   help='Rebase your branch on top of wrap\'s revision (git only)')
+    add_common_arguments(p)
+    p.set_defaults(subprojects_func=update)
+
+    p = subparsers.add_parser('checkout', help='Checkout a branch (git only)')
+    p.add_argument('branch_name', nargs='?',
+                   help='Name of the branch to checkout or create (default: revision set in wrap file)')
+    add_common_arguments(p)
+    p.set_defaults(subprojects_func=checkout)
+
+    p = subparsers.add_parser('download', help='Ensure subprojects are fetched, even if not in use. ' +
+                                               'Already downloaded subprojects are not modified. ' +
+                                               'This can be used to pre-fetch all subprojects and avoid downloads during configure.')
+    add_common_arguments(p)
+    p.set_defaults(subprojects_func=download)
+
+def run(options):
+    src_dir = os.path.relpath(os.path.realpath(options.sourcedir))
+    if not os.path.isfile(os.path.join(src_dir, 'meson.build')):
+        mlog.error('Directory', mlog.bold(src_dir), 'does not seem to be a Meson source directory.')
+        return 1
+    subprojects_dir = os.path.join(src_dir, 'subprojects')
```

Comment:
I think the path to subprojects is still configurable?

Reply:
Yes, but AFAIK it's deprecated for a long time, and I would like this command-line tool to work without a build dir (before configuration) so I cannot know if that option is set to a different value.

Reply:
in mintro we have enough infrastructure to parse this out without a build dir. But it does not actually use it but hardcodes 'subprojects' as well. Maybe we should fix that in a later PR.